[WIP] Alternative renderers

tom95 opened the pull request:
Adds support for alternative renderers and deferred updating. Could be merged as-is but there may be additional changes needed to fully make alternative renderers work and thus marked as work-in-progress.

* drop distinction between keyed- and index children collections in TreeEntry to facilitate detecting move and delete operations (was not a problem in Morphic because it always overrode the entire Morph list)
* NodeTree now carries a renderer property that is a subclass of CmfReact
* reconciliation calls into the renderer, by default these calls are only applied on the element tree; Morphic could potentially also benefit from moving updating there
* add a draft for a godot renderer (no additional dependencies) to illustrate concept

tom95 committed:
adds support for alternative renderers and deferred updating

* drop distinction between keyed- and index children collections in TreeEntry to facilitate detecting move and delete operations (was not a problem in Morphic because it always overrode the entire Morph list)
* NodeTree now carries a renderer property that is a subclass of CmfReact
* reconciliation calls into the renderer, by default these calls are only applied on the element tree; Morphic could potentially also benefit from moving updating there
* add a draft for a godot renderer (no additional dependencies) to illustrate concept

tom95 committed:
Merge branch 'main' of github.com:cmfcmf/ReactS

tom95 commented:
Easiest way to test the Godot render:
```smalltalk

g := CMFReactGodot new.
g initAndRender: ReactGodotExampleComponent asReactNodeBuilder build.
g render.
```
Then change the GodotExampleComponent, call `g render`, and observer the Trascript

cmfcmf commented on packages/Cmfcmf-React-Godot.package/CMFReactNodeGodot.class/instance/asCMFReactNode.st:
```diff
@@ -0,0 +1,4 @@
+converting
+asCMFReactNode
+
+	
```
Should this return `self`?

tom95 replied:
yep SB removes these because the parser adds these when they are missing, so SB takes a guess to always remove them... should fix this -.-

tom95 committed:
add a lifecycle function that wraps each frame update (no-op by default); extend godot examples with more corner cases

tom95 committed:
also pass treeEntry to renderer state (needed in godot to find paths to root for the reconciliation actions); fix update and delete for godot

tom95 committed:
add init route for godot server, add godot example

tom95 committed:
ensure that the treeEntry pointer is set up also for root nodes; more experiments on the godot examples

tom95 committed:
remove godot package

tom95 committed:
fix mutating child pointer; fix useEffect hook using stale closure

tom95 committed:
allow other kinds of classes to provide a useContext-like functionality

tom95 committed:
add rudimentary portal support (no impl for morphic backend)

essentially just relies on the render method of the specific node to add its children to the portal's container instead of the current parent

tom95 committed:
allow batching updates

tom95 committed:
remove assert that became quite costly at 10,000+ nodes

tom95 committed:
replace JsonObject with dedicated ChildPointer object

tom95 committed:
drop TreeNodeEntry>>#id in favor of always passing direct references to entries

* massively improves performance as there is less indirection
* hooks now live in the nodeEntries directly
* effects now run in a separate phase after rendering has completed
* to be investigated: drops hook recovery mechanism after code change

tom95 committed:
proposal: allow instantiating custom components via #new and use #doesNotUnderstand: to assign props; no more build message necessary

tom95 committed:
offer separate init: from initAndRender:

tom95 committed:
fix print strings for incomplete nodes; first steps to fix codeChanged:; convenience method for treating a different method than render: as new component

tom95 committed:
split codeChanged: into a sync and async part for easier consompution by non-morphic backends

cmfcmf commented on packages/Cmfcmf-React-Tests.package/CMFReactRemovesOutdatedHooksTest.class/instance/ConditionalChild..st:
```diff
@@ -1,6 +1,8 @@
-	CMFReact useState: 123.
+	CMFReact useEffect: [
+		"tricky stuff here: we want to just set the instVar; however, when using method components, we actually instantiate a copy of the TestCase class that our React system operates on. So, we go all the way up the stack to find the original instance of our TestCase class."
+		[(thisContext findContextSuchThat: [:context | context selector = #testWorks]) receiver markCalledCleanup]].
```
🙈
